### PR TITLE
 wrappers/services.go: introduce EnsureSnapServices() (2.50)

### PR DESCRIPTION
**Pull request by mvo5**
Backport of https://github.com/snapcore/snapd/pull/10133 for 2.50 - with that we can cherry-pick PR #10164 (commit  406dbc0 from master).

As suggested this now also includes #10151 and  #10171

### CODE REVIEW

File: wrappers/services.go [hunk 1]
```diff
@@ -375,120 +379,270 @@ func userDaemonReload() error {
...
+// entire transaction and to every snap service that is ensured, and options
+// which are per-snap service and specified in the map argument.
+// The return value is a map of the snap's Info the a list of App's that were
+// updated or modified. Apps in a snap that did not have their service
+// definitions change are not included in the map's list value.
+// If any errors are encountered trying to update systemd units, then all
+// changes performed up to that point are rolled back, meaning newly written
+// units are deleted and modified units are attempted to be restored to their
+// previous state. In this case the modified return value is nil.
+func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSnapServicesOptions, inter interacter) (modified map[*snap.Info][]*snap.AppInfo, err error) {
```
Comment (anonymouse64) [inline wrappers/services.go:418]: hmm this still has the modified return value, I would have expected this PR to also have included this PR from @pedronis: https://github.com/snapcore/snapd/pull/10176

Comment (mvo5) [reply wrappers/services.go:463]: Indeed, I updated the PR to include the other two cherry picks here for easier review

Comment (anonymouse64) [reply wrappers/services.go:463]: this PR is still missing https://github.com/snapcore/snapd/commit/e7f02ea82b60149edaa08378956e8be06d9bac5f

**Review by anonymouse64 — APPROVED** [review]
+1

**Pull request merged**